src/com/google/javascript/jscomp/deps/ModuleLoader.java

```diff
@@ -564,4 +607,30 @@ public String apply(DependencyInfo info) {
      */
     NODE
   }
+
+  private static class InputPathData {
+    final String path;
+
+    InputPathData(String path) {
+      this.path = path;
+    }
+
+    boolean isRelativeToCompilationRoot() {
+      return !(isAbsoluteIdentifier(this.path) || isRelativeIdentifier(this.path));
+    }
+
+    String ensureLeadingSlash() {
```

Comment:
The struggle I have is one of terminology. Within this file:

 * isAbsoluteIdentifier('/foo') => true
 * isAbsoluteIdentifier('./foo') => false
 * isAbsoluteIdentifier('foo') => false
 * isRelativeIdentifier('/foo') => false
 * isRelativeIdentifier('./foo') => true
 * isRelativeIdentifier('foo') => false

I don't know how to label paths that do not begin with either a '.' or '/' as they are neither relative or absolute.

If it's a module path `foo/bar.js` can either be relative to the compilation root (LEGACY mode), looked up from the node module registry (NODE mode) or invalid (BROWSER mode).

If it's an input path, `foo/bar.js` is relative to either the compilation or module roots.

It's all horribly confusing.